admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,444 @@
+#define ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_IMPL_HPP_
+
+#include "admittance_controller/admittance_rule.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "rclcpp/duration.hpp"
+#include "rclcpp/utilities.hpp"
+#include "tf2_ros/transform_listener.h"
+
+namespace admittance_controller
+{
+/// Configure admittance rule memory for num joints and load kinematics interface
+controller_interface::return_type AdmittanceRule::configure(
+  const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> & node, const size_t num_joints)
+{
+  num_joints_ = num_joints;
+
+  // initialize ROS functions
+  clock_ = node->get_clock();
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(clock_);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  // initialize memory and values to zero  (non-realtime function)
+  reset(num_joints);
+
+  // Load the differential IK plugin
+  if (!parameters_.kinematics.plugin_name.empty())
+  {
+    try
+    {
+      kinematics_loader_ =
+        std::make_shared<pluginlib::ClassLoader<kinematics_interface::KinematicsInterface>>(
+          parameters_.kinematics.plugin_package, "kinematics_interface::KinematicsInterface");
+      kinematics_ = std::unique_ptr<kinematics_interface::KinematicsInterface>(
+        kinematics_loader_->createUnmanagedInstance(parameters_.kinematics.plugin_name));
+      if (!kinematics_->initialize(
+            node->get_node_parameters_interface(), parameters_.kinematics.tip))
+      {
+        return controller_interface::return_type::ERROR;
+      }
+    }
+    catch (pluginlib::PluginlibException & ex)
+    {
+      RCLCPP_ERROR(
+        rclcpp::get_logger("AdmittanceRule"), "Exception while loading the IK plugin '%s': '%s'",
+        parameters_.kinematics.plugin_name.c_str(), ex.what());
+      return controller_interface::return_type::ERROR;
+    }
+  }
+  else
+  {
+    RCLCPP_ERROR(
+      rclcpp::get_logger("AdmittanceRule"),
+      "A differential IK plugin name was not specified in the config file.");
+    return controller_interface::return_type::ERROR;
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::return_type AdmittanceRule::reset(const size_t num_joints)
+{
+  // reset state message fields
+  state_message_.joint_state.name.assign(num_joints, "");
+  state_message_.joint_state.position.assign(num_joints, 0);
+  state_message_.joint_state.velocity.assign(num_joints, 0);
+  state_message_.joint_state.effort.assign(num_joints, 0);
+  for (size_t i = 0; i < parameters_.joints.size(); ++i)
+  {
+    state_message_.joint_state.name = parameters_.joints;
+  }
+  state_message_.mass.data.resize(6, 0.0);
+  state_message_.selected_axes.data.resize(6, 0);
+  state_message_.damping.data.resize(6, 0);
+  state_message_.stiffness.data.resize(6, 0);
+
+  // reset admittance state
+  admittance_state_ = AdmittanceState(num_joints);
+
+  // reset transforms and rotations
+  trans_ = Transforms();
+  trans_ref_ = Transforms();
```

Comment:
why?

Reply:
I removed the two instances of Transforms in favor of one. I also renamed the struct to `AdmittanceTransforms` and the instance name to `admittance_transforms_`